### ISSUE TITLE
Kraken.js as middleware for another Kraken.js instance

### ISSUE THREAD

**Issue by qlonik** [krakenjs/kraken-js#409]
I found a few examples ([1](https://github.com/krakenjs/kraken-js/wiki/Using-krakenjs-middleware-config-for-express-subapps-mounting), [2](https://github.com/jasisk/middleware-patterns#subapp-pattern)) on how to connect express app as a subapp/middleware to parent kraken app and that does not seem to be a problem.

Since the Kraken.js instance is itself an express app we could connect an instance of kraken to another parent kraken instance. 

I am wondering how I can access the configs of the Kraken parent instance. It seems to be possible to access those settings via attaching `app.on('mount', function(parentApp){})` and accessing kraken instance of parentApp. Is it possible to automatically merge the config of the child app with the config of the parent application, giving priority to the config of the child kraken application?

**Comment by jasisk** [comment]
Hey there. That's an interesting question and one we discussed internally for a long time. Ultimately, we decided that—like vanilla express app instances—"kraken apps" shouldn't be wholly independent from their parent apps. That way, you design your app to work the same whether it's mounted or listening.

That said, express does something [quite handy for you](https://github.com/strongloop/express/blob/2f8ac6726fa20ab5b4a05c112c886752868ac8ce/lib/application.js#L224): it assigns the parent app to a property called, `parent` on your app instance.

Check it out:
``` sh
$ node <<EOF
var express = require('express');
var assert = require('assert');
var app = express();
var app2 = express();
app2.on('mount', function () {
  assert(app === app2.parent);
  console.log('all assertions passed!');
});
app.use('/test', app2);
EOF
all assertions passed!
$
```

So if you _wanted_ to merge the configs, you most certainly could. You could, for example, create a new [confit] config inside on `onconfig`, and merge in both `app.parent.kraken` and `config`, then pass that new confit object in the `onconfig` callback.

[confit]: https://github.com/krakenjs/confit

**Comment by qlonik** [comment]
If I understood you correctly, you are saying that if I pass the reference to an express app into onconfig function, I could access parent application using `app.parent` and therefore access settings of the parent application using `app.parent.kraken`.

Isn't onconfig function called during the startup of kraken middleware and is called before the app is mounted?

So, for example, if I do something like this:
```js
var child = express();
var childConfig = function(app) {
  onconfig: function (config, next) {
    // app.parent is available and links to parent
    next(null, config);
  };
};
child.use(kraken(childConfig(child)));

var parent = express();
var parentConfig = function() {
  onconfig: function (config, next) {
    next(null, config);
  };
};
parent.use(kraken(parentConfig()));
parent.use('/path', child);
```

Since `child.use(kraken)` is called before child is attached to parent, onconfig function of child would execute before parent is attached to express instance, and therefore `app.parent` will be unavailable in the onconfig function. Is it not the case?

**Comment by jasisk** [comment]
Kinda sorta. There's a lot of stuff going on here so I'll try to break it down. Before I do, here is an [example project](https://github.com/jasisk/merged-config-example) demonstrating what you want.

You're right in your assertion that `child.use(kraken)` fires `mount` inside kraken before `mount` fires in `child`. However, beyond that, everything else is subject to synchronicity so there is no guarantee of anything.

Fortunately, there are enough lifecycle events that you can make your own guarantee:

In the `child`, you [can determine if you're main or a sub-app](https://github.com/jasisk/merged-config-example/blob/eeaef8b1348e0eba72d39b8c757fb09abc044e0f/sub-app/index.js#L8) by comparing against [require.main](https://nodejs.org/api/modules.html#modules_accessing_the_main_module).

If you're not the main, you can assume you're being mounted so you can [listen for `mount` on the `child`](https://github.com/jasisk/merged-config-example/blob/eeaef8b1348e0eba72d39b8c757fb09abc044e0f/sub-app/index.js#L15). At this point, you can [start the kraken initialization by `child.use`ing it](https://github.com/jasisk/merged-config-example/blob/eeaef8b1348e0eba72d39b8c757fb09abc044e0f/sub-app/index.js#L18-L27). You can use _this_ `kraken`'s `onconfig` to do [whatever hackery you want](https://github.com/jasisk/merged-config-example/blob/eeaef8b1348e0eba72d39b8c757fb09abc044e0f/sub-app/index.js#L20-L25) (and my example _is_ some hackery :grinning:).

Currently, we **don't** communicate the child `start` back up the parent, but as they're intended to be independent, that should be understandable (and you can easily create your own `child.initChild(cb)` to facilitate that).

Does that clear it up a bit? It gets a little convoluted when you want to forge the relationships but it's certainly possible.

**Comment by jasisk** [comment]
Closing this out but let me know if you have other questions.

**Comment by qlonik** [comment]
Yea, sorry, I had question, but then I got distracted :) 

The main feature that you suggested in the last message is to add initialization (i.e. `child.use(kraken)`) inside the handler of the child's 'on mount' event. Does this reliably lead to `app.parent.kraken` being initialized in the child's onconfig function?

You could add 
```js
parent.on('start', function() {
  parent.use('/path', child);
});
```
And this seem to make it 100% reliable for child, because start event is emitted when all the bootstrapping is done including setting up kraken property.

Another reliable way seems to be is to use meddleware. Since all meddleware modules are loaded after the loading of configuration is done, for all meddleware modules `app.parent.kraken` will also be defined.

**Comment by jasisk** [comment]
The example project I linked to above is completely reliable (all the way down to only `listening` on the parent when the child is ready), for what it's worth.

The problem is that both `parent` and `child` emit `start` asynchronously. There is no guarantee one will emit before the other.

One way to solve that, like you've shown, is serializing them. [That's basically what is being done in the example project](https://github.com/jasisk/merged-config-example/blob/48c898a67020a674e9a3b5d35fd047a73ab24446/sub-app/index.js#L28-L31). If the `child` app is mounted, that's a guarantee the parent app's config us loaded (middleware is mount after the config is resolved). It also ensures `child.parent` is resolved.

Check out the example project and let me know what you think. :grinning:

**Comment by qlonik** [comment]
Your example application is great. Thank you for sharing.

I tried create new merged store in onconfig
```js
  function onconfig(config, next) {
    confit(config._store)
      .addDefault(app.parent.kraken._store)
      .create(next);
  }
```
But I ended up in a recursive loop, where sub-app would attach one more sub-app to itself, which attaches another sub-app and so on. I suppose this is expected behaviour, since adding default store from parent contains settings for attaching another middleware as sub-app.
